### ISSUE TITLE
Exception in callback of async function: TypeError: callback is not a function

### ISSUE THREAD

**Issue by dnish** [cult-of-coders/redis-oplog#138]
Hey,
having an issue with using settings like pushToRedis while executing a remove command. We using it server only for a cron job:

    Messages.remove({userId: id}, {pushToRedis: false});

In dev mode this gives me an:

    Exception in callback of async function: TypeError: callback is not a function

On production we get:

    Exception in callback of async function: RangeError: attempt to write outside buffer bounds
     at Object.Future.wait (/var/app/bundle/programs/server/node_modules/fibers/future.js:446:16)

**Comment by dnish** [comment]
Okay, the error on the client side was my mistake (redis-oplog wasn't initialized), but we still have the issue on the production server. This is the full message:

```
CJ - DELETE OLD USERS AND MESSAGES
Exception in callback of async function: RangeError: attempt to write outside buffer bounds
    at Object.Future.wait (/var/app/bundle/programs/server/node_modules/fibers/future.js:446:16)
    at [object Object].<anonymous> (packages/meteor.js:213:24)
    at [object Object].MongoConnection.(anonymous function) (packages/mongo/mongo_driver.js:778:49)
    at [object Object].mongoConnectionProto.(anonymous function) [as remove] (packages/meteorhacks_kadira.js:3288:32)
    at OptimizedApply (packages/meteorhacks_kadira.js:291:17)
    at Object.ret.(anonymous function) [as remove] (packages/meteorhacks_kadira.js:3246:16)
    at [object Object].remove (packages/mongo/collection.js:616:29)
    at [object Object].remove (packages/cultofcoders:redis-oplog/lib/mongo/Mutator.js:168:45)
    at [object Object]._.extend.remove (packages/cultofcoders:redis-oplog/lib/mongo/collection.extension.js:53:35)
    at server/cronjobs.js:42:14
    - - - - -
    at Buffer.write (buffer.js:638:11)
    at serializeString (/var/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/bson/lib/bson/parser/serializer.js:41:83)
    at serializeInto (/var/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/bson/lib/bson/parser/serializer.js:600:17)
    at serializeObject (/var/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/bson/lib/bson/parser/serializer.js:308:18)
    at serializeInto (/var/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/bson/lib/bson/parser/serializer.js:776:17)
    at serializeObject (/var/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/bson/lib/bson/parser/serializer.js:308:18)
    at serializeInto (/var/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/bson/lib/bson/parser/serializer.js:776:17)
    at serializeObject (/var/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/bson/lib/bson/parser/serializer.js:308:18)
    at serializeInto (/var/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/bson/lib/bson/parser/serializer.js:776:17)
    at serializeObject (/var/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/bson/lib/bson/parser/serializer.js:308:18)
```

Cronjob:

    //Deleting old users

    let d = new Date();
    d.setDate(d.getDate() - 30); //Delete after 30 days inactivity

    let users = Meteor.users.find({'status.lastPing': {$lt: d}}).fetch();
    let userIds = _.pluck(users, '_id');

    userIds.forEach((id) => {
        Messages.remove({userId: id}, {pushToRedis: false});
        Chats.remove({$or: [{user1: id}, {user2: id}]}, {pushToRedis: false});
        Meteor.users.remove({_id: id},{pushToRedis: false});
    });

It's a large remove (>100k docs).

**Comment by theodorDiaconu** [comment]
@dnish 
https://docs.mongodb.com/manual/reference/limits/

It is most likely related to this. I think I know whatsup. When pushToRedis is false, we should default to main behavior. Currently if we do a large remove, first we search them all, then do some stuff... That is the problem here for sure. Great find!

**Comment by theodorDiaconu** [comment]
Fixed both issues. The one with: `Exception in callback of async function: TypeError: callback is not a function` and the limits one in the new PR.

**Comment by dnish** [comment]
@theodorDiaconu Yep, can confirm. Both issues are fixed. Thank you very much ☺️

**Comment by dnish** [comment]
Okay, our app started flickering without using optimistic ui (if a user sends a message, the message if flickering). I'll check out what's going on.

**Comment by dnish** [comment]
Just added a find within Tracker again:

    Tracker.autorun(function() { Messages.find({}).fetch(); console.log('New data');});


Added a new message:

1. Without redis-oplog: 2x executed per new message
2. With redis-oplog: 3x executed per new message

Something seems to be still wrong. I'm using latest 1.2 build.

**Comment by theodorDiaconu** [comment]
@dnish 

Ok, so optimistic-ui will have a small increase to your method response times but it makes sense, let me explain what meteor does to achieve optimistic-ui:

1. Does the mutation in the local collection of the client
2. Sends call to the server to do it in the real database
3. Before the method returns, the server needs to send the correct events based on publication like "changed","added","removed"
4. Method returns
5. Client sees if the changes he made to the local collection match the events sent by the server between 3 - 4. If nothing is sent by the server, the client simply reverts

Extra CPU cost will come from number 3. Because it needs to see what changes to send you based on the publications you've got.

**Comment by theodorDiaconu** [comment]
Optimistic-UI in a sense implies the fact that you would like to wait longer for the server, so that you can have instant updates on UI. In your server-side method, I wouldn't forget to put "this.unblock()" so other methods can still be executed in paralel.

**Comment by dnish** [comment]
Oh okay, I've thought that the server only needs to send back the result like in a server only method and the client is responsible to merge/check the server result.
